src/dom.js

```diff
@@ -0,0 +1,371 @@
+  //   This pretty much is <pre></pre> lifted to be a css property.
+  const style = window.getComputedStyle(node);
+  const inline = Boolean(style['display'].match(/^(content|inline.*|)$/));
+  const preformatted = Boolean(style['white-space'].match(/^pre.*/));
+
+  // This is where the actual space collapsing happens
+  const collapseTextAcrossInlineNodes = () => {
+    let prev;
+    let insertSpace = inline && preformatted;
+    // We collected a list of all textNodes/pre formatted elements across inline elements
+    // in the current box down below, now we can apply cross element
+    // space collapsing
+    each(enumerate(inlineTextNodes), ([idx, elm]) => {
+      const isText = elm.nodeName === '#text';
+      const val = isText ? elm.nodeValue : undefined;
+
+      insertSpace = insertSpace || Boolean(isText && val.match(/^\s/));
+
+      // Removal of text nodes consisting solely of white space
+      if (isText && val.matches(/^\s*$/)) {
+        elm.remove();
+      }
+
+      // This collapses space inside a single textNode..
+      const insertSpaceNext = isText && val.match(/\s$/);
+      if (isText) {
+        elm.nodeValue = elm.nodeValue.trim().replace(/\s+/g, ' ');
+      }
+
+      // This is what applies our cross element collapsing
+      // NOTE: An implementation just collapsing cross element
+      // spaces to the right would be simpler and still correct
+      // in terms of equivalence; it would also be much less
+      // readable when inspected by humans; this is not very pretty:
+      // `Hello <b>World</b>` -> `Hello<b> World</b>`
+      // `<b>Hello</b> <em>World</em>` -> `<b>Hello</b><em> World</em>`
+      if (insertSpace && prev !== undefined) {
+        const prevParents = Array.from(domParentNodes(prev));
+        const elmParents = Array.from(domParentNodes(elm));
+
+        while (prevParents[0].isSameNode(elmParents[0])) {
+          prevParents.shift();
+          elmParents.shift();
+        }
+        const commonParentNode = prevParents[0];
+
+        if (prev.nodeName === '#text' && prevParents.length === 0) {
+          // Like in the `Hello <b>World</b>` case
+          prev.nodeValue += ' ';
+        } else if (isText && elmParents.length === 0) {
+          // Like in the `<b>Hello</b> World` case
+          elm.nodeValue = ' ' + elm.nodeValue;
+        } else {
+          // Like in the `<b>Hello</b> <em>World</em>` case
+          const before = elmParents[0];
+          const commonParent = before.parentNode;
+          const nu = window.document.createTextNode(' ');
+          commonParent.insertBefore(nu, before);
+        }
+      }
+
+      prev = elm;
+      insertSpace = insertSpaceNext;
+    });
+
+    // clear the array; we cannot just reassign the array
+    // because we need to propagate the change to our parent
+    // calls further up in the recursion chain
+    while (inlineTextNodes.length > 0)
+      inlineTextNodes.pop();
```

Comment:
- probably not the fastest way. consider `inlineTextNodes.legnth = 0`;
- please use `{ }` even for 1 liners.

Reply:
you mean `inlineTextNodes.legnth === 0`?

Reply:
no `= 0`. to clear the array.